zebra-state/src/service/non_finalized_state.rs

```diff
@@ -264,6 +271,49 @@ impl NonFinalizedState {
         Ok(())
     }
 
+    /// Invalidate block with has `block_hash` and all descendants from the non-finalized state. Insert
+    /// the new chain into the chain_set and discard the previous.
+    pub fn invalidate_block(&mut self, block_hash: Hash) {
+        if !self.any_chain_contains(&block_hash) {
+            return;
+        }
+
+        let mut chain = self
+            .find_chain(|chain| chain.contains_block_hash(block_hash))
+            .expect("block hash exist in a chain");
+
+        let new_chain = Arc::make_mut(&mut chain);
+        let block_height = new_chain.height_by_hash(block_hash).unwrap();
+
+        // Split the new chain at the the `block_hash` and invalidate the block with the
+        // block_hash along with the block's descendants
+        let mut invalidated_blocks = new_chain.blocks.split_off(&block_height);
+        for (_, ctx_block) in invalidated_blocks.iter().rev() {
+            new_chain.revert_chain_with(ctx_block, chain::RevertPosition::Tip);
+        }
+
+        self.invalidated_blocks.insert(
+            block_hash,
+            InvalidatedBlockData {
+                block: invalidated_blocks.remove(&block_height).unwrap(),
+                descendants: invalidated_blocks,
+                timestamp: SystemTime::now(),
+            },
+        );
+
+        // If the new chain still contains blocks:
+        // - add the new chain fork or updated chain to the set of recent chains
+        // - remove the chain containing the hash of the block from the chain set
+        if !new_chain.is_empty() {
+            self.insert_with(Arc::new(new_chain.clone()), |chain_set| {
+                chain_set.retain(|c| !c.contains_block_hash(block_hash))
+            });
+
+            self.update_metrics_for_chains();
+            self.update_metrics_bars();
+        }
```

Comment:
It should update the metrics whenever `chain_set` is modified.

Optional/Nitpick: This method could drop the chain and return early when the invalidated block is the non-finalized chain root block.

```suggestion
        }

        self.update_metrics_for_chains();
        self.update_metrics_bars();
```